plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,849 @@
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	gsize fw_size = 0;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 2, "prepare");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 85, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 10, "commit");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 2, "verify");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 1, "reboot");
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* prepare the device to accept firmware image */
+	fw_size = g_bytes_get_size(fw);
+	if (!fu_mediatek_scaler_device_prepare_update(device, fw_size, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* write firmware to device */
+	chunks = fu_chunk_array_new_from_bytes(fw, 0x00, 0x00, DDC_DATA_PAGE_SIZE);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+
+		/* fast forward if chunk is empty, on the contrary set data per fragment size */
+		if (fu_mediatek_scaler_device_chunk_data_is_blank(chk)) {
+			if (!fu_mediatek_scaler_device_set_data_fast_forward(self, chk, error))
+				return FALSE;
+		} else {
+			if (!fu_mediatek_scaler_device_set_data(self, chk, error))
+				return FALSE;
+		}
+
+		/* update progress */
+		fu_progress_set_percentage_full(fu_progress_get_child(progress),
+						(gsize)i + 1,
+						(gsize)chunks->len);
+	}
+
+	if (!fu_mediatek_scaler_device_check_sent_info(self, fw, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* send ISP command to commit the update */
+        if(!fu_mediatek_scaler_device_commit_firmware(self, fw, error))
+                return FALSE;
+	fu_progress_step_done(progress);
+
+	/* verify display and ISP status result */
+	if (!fu_mediatek_scaler_device_verify(device, fw_size, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+        /* reboot the device */
+	if (!(fu_mediatek_scaler_device_set_isp_reboot(self, error)))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_mediatek_scaler_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 98, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 2, "reload");
+}
+
+static void
+fu_mediatek_scaler_device_init(FuMediatekScalerDevice *self)
+{
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INTERNAL);
```

Comment:
I know it is for your initial device; but will this always be true?  If not; any way you can get a hint if it's internal/external?

Reply:
will circle back here to modify the filter with appropriate identifier when an external opportunity comes to the table.